skimage/morphology/_skeletonize_3d_cy.pyx.in

```diff
@@ -158,27 +167,35 @@ cdef void find_simple_point_candidates(pixel_type[:, :, ::1] img,
                 if img[p, r, c] != 1:
                     continue
 
-                is_border_pt = (curr_border == 1 and img[p, r, c-1] == 0 or  #N
-                                curr_border == 2 and img[p, r, c+1] == 0 or  #S
-                                curr_border == 3 and img[p, r+1, c] == 0 or  #E
-                                curr_border == 4 and img[p, r-1, c] == 0 or  #W
-                                curr_border == 5 and img[p+1, r, c] == 0 or  #U
-                                curr_border == 6 and img[p-1, r, c] == 0)    #B
+                # border checking order is 1=N, 2=S, 3=E, 4=W, 5=U, 6=B
+                is_border_pt = (curr_border == 1 and img[p, r, c-1] == 0 or
+                                curr_border == 2 and img[p, r, c+1] == 0 or
+                                curr_border == 3 and img[p, r+1, c] == 0 or
+                                curr_border == 4 and img[p, r-1, c] == 0 or
+                                curr_border == 5 and img[p+1, r, c] == 0 or
+                                curr_border == 6 and img[p-1, r, c] == 0)
                 if not is_border_pt:
                     # current point is not deletable
                     continue
 
                 get_neighborhood(img, p, r, c, neighborhood)
 
                 # check if (p, r, c) can be deleted:
-                # * it must not be an endpoint;
                 # * it must be Euler invariant (condition 1 in [Lee94]_); and
                 # * it must be simple (i.e., its deletion does not change
-                #   connectivity in the 3x3x3 neighborhood)
-                #   this is conditions 2 and 3 in [Lee94]_
-                if (is_endpoint(neighborhood) or
-                    not is_Euler_invariant(neighborhood, Euler_LUT) or
-                    not is_simple_point(neighborhood)):
+                #   connectivity in the 3x3x3 neighborhood; conditions 2 and
+                #   3 in [Lee94]_); and
+                # * it must not be an endpoint for medial axis thinning or
+                #   a surface endpoint for medial surface thinning
+                #   (condition 4 in [Lee94]_)
+                if (not is_Euler_invariant(neighborhood, Euler_LUT) or
+                        not is_simple_point(neighborhood)):
+                    continue
+                if (surface_flag == 0 and
+                        is_endpoint(neighborhood)):
+                    continue
+                elif (surface_flag == 1 and
+                      is_surface_endpoint(neighborhood)):
                     continue
```

Comment:
Just a thought: are any of these condition checks are substantially cheaper or more likely to succeed than the others? If so, perhaps they should occur first so that we `continue` as early as possible before evaluating the other conditions.

Reply:
You're right, this can be improved slightly. I've just ran some basic benchmarking with the conditions in different orders. The changes have been committed along with the benchmark.

Note that this also changes the ordering for the medial axis thinning. However, the new ordering also slightly improves the medial axis thinning, given the particular case used in the benchmark at least.